ceno_zkvm/src/instructions/riscv/div.rs

```diff
@@ -1,123 +1,348 @@
-        let outcome_value = outcome.value();
+        // The soundness analysis for these constraints is only valid for
+        // 32-bit registers represented over the Goldilocks field, so verify
+        // these parameters
+        assert_eq!(UInt::<E>::TOTAL_BITS, u32::BITS as usize);
+        assert_eq!(E::BaseField::MODULUS_U64, goldilocks::MODULUS);
+
+        // 32-bit value from rs1
+        let dividend = UInt::new_unchecked(|| "dividend", cb)?;
+        // 32-bit value from rs2
+        let divisor = UInt::new_unchecked(|| "divisor", cb)?;
+        let quotient = UInt::new(|| "quotient", cb)?;
+        let remainder = UInt::new(|| "remainder", cb)?;
+
+        // `rem_e` and `div_e` are expressions representing the remainder and
+        // divisor from the signed or unsigned division operation, with signs
+        // normalized to be nonnegative, so that correct values must satisfy
+        // either `0 <= rem_e < div_e` or `div_e == 0`.  The `rem_e` value
+        // should be constrained to be nonnegative before being returned from
+        // this block, while the checks `rem_e < div_e` or `div_e == 0` are
+        // done later.
+        let (internal_config, rem_e, div_e) = match I::INST_KIND {
+            InsnKind::DIVU | InsnKind::REMU => {
+                cb.require_equal(
+                    || "unsigned_division_relation",
+                    dividend.value(),
+                    divisor.value() * quotient.value() + remainder.value(),
+                )?;
+
+                (InternalDivRem::Unsigned, remainder.value(), divisor.value())
+            }
+
+            InsnKind::DIV | InsnKind::REM => {
+                let dividend_signed =
+                    Signed::construct_circuit(cb, || "dividend_signed", &dividend)?;
+                let divisor_signed = Signed::construct_circuit(cb, || "divisor_signed", &divisor)?;
+                let quotient_signed =
+                    Signed::construct_circuit(cb, || "quotient_signed", &quotient)?;
+                let remainder_signed =
+                    Signed::construct_circuit(cb, || "remainder_signed", &remainder)?;
+
+                // The quotient and remainder can be interpreted as non-positive
+                // values when exactly one of dividend and divisor is negative
+                let neg_div_expr = {
+                    let a_neg = dividend_signed.is_negative.expr();
+                    let b_neg = divisor_signed.is_negative.expr();
+                    // a_neg xor b_neg
+                    &a_neg + &b_neg - 2 * a_neg * b_neg
+                };
+                let negative_division = cb.flatten_expr(|| "neg_division", neg_div_expr)?;
+
+                // Check for signed division overflow: i32::MIN / -1
+                let is_dividend_max_negative = IsEqualConfig::construct_circuit(
+                    cb,
+                    || "is_dividend_max_negative",
```

Comment:
is_dividend_`min`_negative to match with i32::MIN syntatically

Reply:
Agreed the "max negative" phrasing is a bit awkward -- is `is_dividend_signed_min` okay?